### ISSUE TITLE
"sudo modprobe" fails on NixOS, wrongly tries to access /lib/modules

### ISSUE THREAD

**Issue by bjornfor** [NixOS/nixpkgs#5492]
Plain modprobe looks up the module dir by using the environment variable $MODULE_DIR, which typically points to ```/run/current-system/kernel-modules/lib/modules```. But sudo drops environment variables (default setting) so that modprobe falls back to using /lib/modules... and fails.

https://github.com/NixOS/nixpkgs/pull/5466 suggested to fix this by making sudo preserve $MODULE_DIR, but it was voted down due to security concerns.

The real (safe) fix seems to be modifying the modprobe patch so that it tries the following paths, picking the first that exists: ```$MODULE_DIR```, ```/lib/modules```, ```/run/current-system/kernel-modules/lib/modules``` (or "booted-system"?)

**Comment by kirelagin** [comment]
Khm, #2692 was implemented specifically for this kind of things. `MODULE_DIR` should go to `sessionVariables` instead of [`variables`](https://github.com/NixOS/nixpkgs/blob/a6673133efda088c44d73f0c1ae752f59587e3f4/nixos/modules/system/boot/modprobe.nix#L97), it’s that simple.

**Comment by kirelagin** [comment]
By the way, it looks like `modprobe` [is wrapped](https://github.com/NixOS/nixpkgs/blob/a6673133efda088c44d73f0c1ae752f59587e3f4/nixos/modules/system/boot/modprobe.nix#L13) so I can’t see how can it possibly look for modules in a wrong directory…

**Comment by bjornfor** [comment]
@kirelagin: Interesting. I will do a test with sessionVariables soon.

Ah, there is a wrapped modprobe? I see it now, but it seems the non-wrapped modprobe is first in PATH:

```
$ which -a modprobe
/run/current-system/sw/bin/modprobe
/run/current-system/sw/sbin/modprobe
```

I checked my configuration.nix and I didn't find anything suspicious, so I guess every NixOS system has this? The "bin" version is not wrapped, the "sbin" version is.

Explicitly calling the "sbin" version (the wrapped one) doesn't break with sudo, as one would expect since it hardcodes $MODULE_DIR.

**Comment by bjornfor** [comment]
Ah, maybe this is a regression since we merged sbin into bin? ab04b7d0bba24d068ee57c25050ecf73a428c552

**Comment by kirelagin** [comment]
Very likely.

I guess if `$MODULE_DIR` is moved to `sessionVariables` the wrapper will not be needed any more (and systemd units have `MODULE_DIR` set explicitly anyway).

**Comment by puffnfresh** [comment]
Ah, thanks for this issue. I kept trying to modprobe a module but it would always fail. Got it working by just letting environment variables through:

```bash
sudo -E modprobe v4l2loopback
```

:fireworks:

**Comment by basvandijk** [comment]
I also just tripped on this.

NixOS's `strongSwan` module needs to use `modprobe` so it [puts](https://github.com/NixOS/nixpkgs/blob/4bb1e611a2ac12ceb309566e44a172149aeb7539/nixos/modules/services/networking/strongswan.nix#L121) `kmod` in the `PATH`. I noticed errors in the journal saying that certain kernel modules weren't loaded. When I replaced `kmod` with `config.system.sbin.modprobe` the errors disappeared.
